### PR TITLE
Run keycloakify behind corporate proxy

### LINKED ISSUE #257: Run keycloakify behind corporate proxy
If I run keycloakify behind a corporate proxy, I get the following error:

```
$ npx keycloakify
🔏 Building the keycloak theme...⌚
node:events:491
      throw er; // Unhandled 'error' event
      ^

Error: connect ETIMEDOUT 140.82.121.3:443
    at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1300:16)
Emitted 'error' event on ClientRequest instance at:
    at TLSSocket.socketErrorListener (node:_http_client:494:9)
    at TLSSocket.emit (node:events:513:28)
    at emitErrorNT (node:internal/streams/destroy:151:8)
    at emitErrorCloseNT (node:internal/streams/destroy:116:3)
    at process.processTicksAndRejections (node:internal/process/task_queues:82:21) {
  errno: -4039,
  code: 'ETIMEDOUT',
  syscall: 'connect',
  address: '140.82.121.3',
  port: 443
}

Node.js v18.11.0
```

### Note

The npm config vars `https-proxy`, `no-proxy` and `proxy` are set and work well for the dependency installation and within other projects.

### PR DESCRIPTION
**Pull request by lordvlad**
Fixes #257

Use make-fetch-happen for the download step. This lib will use `PROXY`
and `HTTPS_PROXY` and `NO_PROXY` env vars out of the box.

Additionally we'll try and get proxy config from npm. Unfortunately,
the most straightforward options is to call npm config to do this, since
npm  config is not easily extracted as a lib and we don't want to
replicate the resolution mechanisms.

### CODE REVIEW

**Comment by lordvlad** [comment]
unfortunately, the newly introduced lib wants to  run node LTS only :(

**Comment by lordvlad** [comment]
@garronej you _could_ disable the `engines` check https://stackoverflow.com/a/57748163. Depending on what level of support you want to guarantee users. Personally, I'd drop v15 and v17 from the CI pipeline, and introduce v18 instead. WDYT?

**Comment by garronej** [comment]
Wow, great !  

Yes let's drop v15 and 17 from the matrix

**Comment by garronej** [comment]
@lordvlad Fantastic, all green.  
Next time would you create your next PR from a branch this way we could release a candidate just by updating to `4.12.6-rc.0` for example.  
Otherwise it look good to me, is there something to document? Do you want to release a candidate or release directly?

**Comment by lordvlad** [comment]
> Next time would you create your next PR from a branch this way we could release a candidate just by updating to 4.12.6-rc.0 for example.

@garronej  Not sure I get it. I don't see any RC branches, I could branch from. For the PR I did create a branch `lordvlad/issue257` (used VS Code's GitHub integration, wanted to try it out)

**Comment by lordvlad** [comment]
> is there something to document? Do you want to release a candidate or release directly?

I'd wait  for feedback, as I was only able to test it in a very small setup on localhost.

**Comment by garronej** [comment]
> Not sure I get it. I don't see any RC branches, I could branch from. For the PR I did create a branch `lordvlad/issue257` (used VS Code's GitHub integration, wanted to try it out)  

Sorry, my bad again, it's a branch indeed!   

Let's release a candidate then

**Comment by Mstrodl** [comment]
@lordvlad I'm flattered, but I'm not a maintainer :P
image

File: src/bin/tools/downloadAndUnzip.ts [hunk 1]
```diff
@@ -32,69 +27,57 @@ async function maybeStat(path: string) {
...
  * @param dir target directory
+ * @param filename target filename
  * @returns promise for the full path of the downloaded file
  */
-async function download(url: string, dir: string): Promise<string> {
-    await mkdir(dir, { recursive: true });
-    const filename = pathBasename(url);
+async function download(url: string, dir: string, filename: string): Promise<string> {
+    const proxyOpts = await getNpmProxyConfig();
+    const opts: FetchOptions = { cachePath: join(homedir(), ".keycloakify/cache"), ...proxyOpts };
```
Comment (Mstrodl) [inline src/bin/tools/downloadAndUnzip.ts:75]: Consider respecting `XDG_CACHE_HOME`

**Review by Mstrodl — APPROVED** [review]
Looks sane to me

**Comment by lordvlad** [comment]
Haha GitHub suggested you and I just rolled with it 😅

File: src/bin/tools/downloadAndUnzip.ts (hunk 1 above)
Comment (lordvlad) [reply src/bin/tools/downloadAndUnzip.ts:75]: good point :)

**Pull request merged**